### ISSUE TITLE
Cluster Error Confusing

### ISSUE THREAD

**Issue by kempy007** [cockroachdb/cockroach#111353]
Following Guide Step1 at https://www.cockroachlabs.com/docs/stable/start-a-local-cluster

My steps were:
```
cockroach start --http-addr=localhost:8090 --listen-addr=localhost:26258 --insecure --store=node1 --join=localhost:26259,localhost:26258

cockroach start --http-addr=localhost:8091 --listen-addr=localhost:26259 --insecure --store=node2 --join=localhost:26258,localhost:26259

```

I keep getting this error.

'''
*
* WARNING: The server appears to be unable to contact the other nodes in the cluster. Please try:
*
* - starting the other nodes, if you haven't already;
* - double-checking that the '--join' and '--listen'/'--advertise' flags are set up correctly;
* - running the 'cockroach init' command if you are trying to initialize a new cluster.
*
* If problems persist, please see https://www.cockroachlabs.com/docs/v23.1/cluster-setup-troubleshooting.html.
*
'''

However If I ignore this and run 'cockroach init --host=localhost:26258 --insecure' It works as expected.

Jira issue: CRDB-31869

**Comment by yuzefovich** [comment]
Can you share your exact reproduction steps?

**Comment by yuzefovich** [comment]
You do have to run `cockroach init` to initialize the cluster - this is sub-step 4 from Step 1 in the guide. Until you haven't run `cockroach init`, it's expected to see the message on the nodes. In fact, that message says "running the 'cockroach init' command if you are trying to initialize a new cluster." as the suggestion too.

**Comment by kempy007** [comment]
My point is the message is not true 'WARNING: The server appears to be unable to contact the other nodes in the cluster.' It should reflect that the cluster has not been initialised, can you not differentiate between these conditions?

**Comment by knz** [comment]
We can't differentiate between these two conditions. However, we can make the warning include the phrase "unable to contact other nodes OR the cluster is not initialized yet". Would that work?

**Comment by kempy007** [comment]
So I think I would want to add a check that the ports are open using the -join= inputs and append a status message of the ports being reachable or not, that would make it less confusing if this is a missing config(absolutely true before init) or network issue(potentially true).

I had a quick look and the message seems split across these two files.
pkg/cli/init.go L98 "Warning:" 
pkg/cli/start.go L595 "The server appears to be unable to contact ...."

Trying to find where -join flag is consumed I see;
L188 pkg/server/init.go
L172 pkg/roachprod/install/cockroach.go
L2337 pkg/server/node.go

I haven't traced through it yet!

Any tips on where I should have a crack at this??

**Comment by knz** [comment]
> add a check that the ports are open using the -join= inputs and append a status message of the ports being reachable or not

This sentence sounds easier and much easier than the complexity of implementation. To be honest we wouldn't even know what a good networking check would look like. Welcoming suggestions before we even talk about code.

**Comment by kempy007** [comment]
As a hacker I would make a connection to the port, if it does not get dropped right away I could assume it is open, taking it further I may send some data to see if a response is sent that I can use to identify the service. Some services will make this easier by sending a banner upon successful connection. 

So if I do 'nc localhost 26257'  I will get a what seems like a hung prompt. if I type ola and press return, it sends back @������ So this is a good test to check the port is open and gives me an idea by the behaviour that it is cockroachdb service. 
Thats pretty much what nmap or similar tools would do, but blast the port with thousands of patterns, and do this by sweeping IP space and port ranges, to map out a network and identify the ports with a service.

In this case you know the hosts and ports and the service and its behaviour.  

Sorry...now talking code!

[Reading the postgressql spec](https://www.postgresql.org/docs/16/protocol-flow.html#PROTOCOL-FLOW-START-UP) I see that it expects a startup message, possibly binary mode, can't see anything related to encoding. I've made an assumption CRDB is 80% similar. 

I found an [example of this message here](https://stackoverflow.com/questions/35669615/how-to-start-postgresql-message-flow-protocol-using-node-js-net-socket) in nodejs. 

This diagram shows the expected payloads, and [has more detail here](https://info.deal-corner.com/post/311c1057-f94b-4b48-aadc-02de4ee5f0e0)

![Protocol Flow](https://info.deal-corner.com/blogs/pg_message_formats.png)

So to summarise what makes a good health check.
- Make a TCP Socket connection
- Send startup message
- Check response is as expected
- Close the connection

I have not considered how transport encryption would alter this check.
